File: day_5_next_permutation.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    void display(vector<int> &A) {
        int n = A.size();
        for (int i = 0; i < n; i++) {
            cout << A[i] << " ";
        }
        cout << endl;
    }

    void nextPermutation(vector<int> &A) {
        int n = A.size();

        int ind = -1;
        for (int i = n - 2; i >= 0; i--) {
            if (A[i] < A[i + 1]) {
                ind = i;
                break;
            }
        }

        if (ind == -1) {
            reverse(A.begin(), A.end());
            return;
        }

        for (int i = n - 1; i > ind; i--) {
            if (A[i] > A[ind]) {
                swap(A[i], A[ind]);
                break;
            }
        }
        
        reverse(A.begin() + ind + 1, A.end());
    }
};

int main() {
    int n;
    cin >> n;
    vector<int> arr;

    for (int i = 0; i < n; i++) {
        int x;
        cin >> x;
        arr.push_back(x);
    }

    Solution solution;
    solution.nextPermutation(arr);
    solution.display(arr);

    return 0;
}

File: Day5NextPermutation.java
import java.util.Scanner;

public class Day5NextPermutation {
    private class Solution {
        private void reverse(int[] nums, int start, int end) {
            while(start < end) {
                int temp = nums[start];
                nums[start] = nums[end];
                nums[end] = temp;
                start++;
                end--;
            }
        }

        private void swap(int[] nums, int num1, int num2) {
            int temp = nums[num1];
            nums[num1] = nums[num2];
            nums[num2] = temp;
        }

        void nextPermutation(int[] nums) {
            int n = nums.length;

            // Step 1: Find the break point:
            int ind = -1; // break point
            for(int i = n-2; i >= 0; i--) {
                if(nums[i] < nums[i+1]) {
                    // index i is the break point
                    ind = i;
                    break;
                }
            }

            // If break point does not exist:
            if(ind == -1) {
                // reverse the whole array:
                reverse(nums, 0, n-1);
                return;
            }

            // Step 2: Find the next greater element
            //         and swap it with nums[ind]:
            for(int i = n-1; i > ind; i--) {
                if(nums[i] > nums[ind]) {
                    swap(nums, i, ind);
                    break;
                }
            }

            // Step 3: reverse the right half:
            reverse(nums, ind+1, n-1);
        }
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int[] nums = new int[n];
        
        for(int i = 0; i < n; i++) {
            nums[i] = scanner.nextInt();
        }

        Solution sol;
        sol = new Day5NextPermutation().new Solution();
        sol.nextPermutation(nums);

        for(int num : nums) {
            System.out.print(num + " ");
        }

        scanner.close();
    }
}
